scripts/notebook-instance-monitor/notebookapi.py

```diff
@@ -0,0 +1,229 @@
+    with open(log_path, 'r') as logs:
+        _logs = json.load(logs)
+    return _logs['ResourceName']
+
+
+def get_notebook_resource_arn():
+    log_path = '/opt/ml/metadata/resource-metadata.json'
+    with open(log_path, 'r') as logs:
+        _logs = json.load(logs)
+    return _logs['ResourceArn']
+
+
+# When is a notebook considered idle by the Notebooks API? - https://github.com/jupyter/notebook/issues/4634
+
+# The way it works at present is that the kernel sends a 'busy' message when it starts executing a request, and an 'idle' message when it finishes. So it's idle if there's not code running. The 'while True' loop would leave it busy.
+
+# Code execution isn't the only kind of request, though. Among other things, when you open a notebook in a tab, it will make a kernel info request, which will reset the timer.
+
+
+def is_idle(last_activity):
+    last_activity = datetime.strptime(last_activity,"%Y-%m-%dT%H:%M:%S.%fz")
+    if (datetime.now() - last_activity).total_seconds() > time:
+        print('Notebook is idle. Last activity time = ', last_activity)
+        return True
+    else:
+        print('Notebook is not idle. Last activity time = ', last_activity)
+        return False
+
+
+response = requests.get('https://localhost:'+port+'/api/sessions', verify=False)
+data = response.json()
+print(data)
+if len(data) > 0:
+    
+    print("Using Jupyter Notebook API since request was successful")
+    for notebook in data:
+
+        if notebook['kernel']['execution_state'] == 'idle':
+            if not ignore_connections:
+                if notebook['kernel']['connections'] == 0:
+                    if not is_idle(notebook['kernel']['last_activity']):
+                        idle = False
+                else:
+                    idle = False #If any connection exists, notebook is not idling
+            else:
+                if not is_idle(notebook['kernel']['last_activity']):
+                    idle = False #If last activity is recent, notebook is not idling
+        else:
+            print('Notebook is not idle:', notebook['kernel']['execution_state'])
+            idle = False
+else:
+    print("Using SageMaker instance last modified time")
+    client = boto3.client('sagemaker')
+    uptime = client.describe_notebook_instance(
+        NotebookInstanceName=get_notebook_name()
+    )['LastModifiedTime']
+    if not is_idle(uptime.strftime("%Y-%m-%dT%H:%M:%S.%fz")):
+        idle = False
+        
+        
+#CPU, Mem and GPU utilization
+print(f"Utilization metrics at {datetime.now()}")
+
+total_cpu_util = psutil.cpu_percent()
+total_mem_util = psutil.virtual_memory().percent
+print(f"CPU utilization = {total_cpu_util}%")
+print(f"Memory utilization = {total_mem_util}%")
+# Testing the GPUtil library for both GPU performance details
+num_gpu = 0
+try:
+    print("GPU utilization = ")
+    with Capturing() as output:
+        GPUtil.showUtilization()
+    
+#     print(output)
```

Comment:
Let's remove commented out code.

Reply:
removed in [this](https://github.com/w601sxs/amazon-sagemaker-notebook-instance-lifecycle-config-samples/commit/855f56118f836f002bcacdcf7e9be93320bfc30d)